server/src-lib/Hasura/GraphQL/Parser/Schema.hs

```diff
@@ -0,0 +1,684 @@
+eqType :: Type k1 -> Type k2 -> Bool
+eqType (NonNullable a) (NonNullable b) = eqNonNullableType a b
+eqType (Nullable    a) (Nullable    b) = eqNonNullableType a b
+eqType _               _               = False
+
+instance HasName (Type k) where
+  getName = getName . discardNullability
+
+instance HasDefinition (Type k) (TypeInfo k) where
+  definitionLens f (NonNullable t) = NonNullable <$> definitionLens f t
+  definitionLens f (Nullable t)    = Nullable <$> definitionLens f t
+
+discardNullability :: Type k -> NonNullableType k
+discardNullability (NonNullable t) = t
+discardNullability (Nullable t)    = t
+
+nullableType :: Type k -> Type k
+nullableType = Nullable . discardNullability
+
+nonNullableType :: Type k -> Type k
+nonNullableType = NonNullable . discardNullability
+
+data NonNullableType k
+  = TNamed (Definition (TypeInfo k))
+  | TList (Type k)
+
+instance Eq (NonNullableType k) where
+  (==) = eqNonNullableType
+
+toGraphQLType :: Type k -> GType
+toGraphQLType = \case
+  NonNullable t -> translateWith False t
+  Nullable    t -> translateWith True  t
+  where
+    translateWith nullability = \case
+      TNamed typeInfo -> TypeNamed (Nullability nullability) $ getName typeInfo
+      TList  typeInfo -> TypeList  (Nullability nullability) $ toGraphQLType typeInfo
+
+
+-- | Like '==', but can compare 'NonNullableType's of different kinds.
+eqNonNullableType :: NonNullableType k1 -> NonNullableType k2 -> Bool
+eqNonNullableType (TNamed a) (TNamed b) = liftEq eqTypeInfo a b
+eqNonNullableType (TList  a) (TList  b) = eqType a b
+eqNonNullableType _          _          = False
+
+instance HasName (NonNullableType k) where
+  getName (TNamed definition) = getName definition
+  getName (TList t)           = getName t
+
+instance HasDefinition (NonNullableType k) (TypeInfo k) where
+  definitionLens f (TNamed definition) = TNamed <$> f definition
+  definitionLens f (TList t)           = TList <$> definitionLens f t
+
+data InputObjectInfo = InputObjectInfo ~[Definition InputFieldInfo]
+-- Note that we can't check for equality of the fields since there may be
+-- circularity. So we rather check for equality of names.
+instance Eq InputObjectInfo where
+  InputObjectInfo fields1 == InputObjectInfo fields2
+    =  Set.fromList (fmap dName fields1)     == Set.fromList (fmap dName fields2)
+data ObjectInfo = ObjectInfo ~[Definition FieldInfo] ~[Definition InterfaceInfo]
+-- Note that we can't check for equality of the fields and the interfaces since
+-- there may be circularity. So we rather check for equality of names.
+instance Eq ObjectInfo where
+  ObjectInfo fields1 interfaces1 == ObjectInfo fields2 interfaces2
+    =  Set.fromList (fmap dName fields1    ) == Set.fromList (fmap dName fields2    )
+    && Set.fromList (fmap dName interfaces1) == Set.fromList (fmap dName interfaces2)
+data InterfaceInfo = InterfaceInfo ~[Definition FieldInfo] ~[Definition InterfaceInfo] ~[Definition ObjectInfo]
```

Comment:
I’d like to see a `Note` that gives an overview of the way objects, interfaces, and unions all fit together. As its stands, this is pretty opaque to me.

Reply:
As a first step towards this, I've made the `*Info` types into records with field names, and added comments documenting their purpose.

Reply:
As both Alexis and me have added some information here, and in particular since Alexis wrote a `Note`, I'm taking the liberty to resolve this item.